cilium/CFP-22215-mutual-auth-for-service-mesh.md

```diff
@@ -0,0 +1,469 @@
+
+Extended agent downtime could impact the ability to properly terminate session authentication for peers, depending on the datapath implementation. N-tuple match on connection endpoints in new auth map discusses how timestamps could be integrated into the datapath to ensure connection termination when mutual authentication reauthentication periods elapse. If these timestamps are not integrated into the datapath, then we would have to accept that authenticated sessions may remain authenticated indefinitely during agent downtime.
+
+### Key question: How to enable and configure mTLS mode
+
+#### Considered option: Define mesh-auth-mode via daemon config
+
+##### Pros
+* Upper bound for auth connections / node is pre-defined
+##### Cons
+* More difficult to test (can't use a always-yes or always-no certificate provider)
+
+#### Option 2: Define mesh-auth-mode via CRD policy
+
+##### Pros
+More user control over the granularity. For high-connection pods like Ingress, it could help to bound the maximum number of auth connections per the discussion under Impact: Authentication connections per node.
+##### Cons
+More complicated
+Let's say two peers have two different configurations, one with node authentication and another with namespace. Does that mean that 2+ different auth sessions are established to provide the authentication the user expects?
+
+### Key question: How to implement dataplane auth check
+
+The dataplane authentication check in eBPF requires two components: One is a bit in the policy logic that marks that affected connections must be authenticated. This can be implemented with minimal changes by adding it into the existing policymap. The second is how we check that a connection is authenticated. This part must be informed by the userspace local control plane populating some state that is available to the eBPF dataplane.
+
+#### Chosen alternative: N-tuple match on connection endpoints in new auth map
+
+A new map is introduced into the eBPF dataplane that allows the packet-handling logic to look up for some subset of `{src, dst, …}` fields of the packet and confirm whether authentication has already occurred for this set of peers.
+
+*Pros*
+- Allow multiple connections between the same peers to leverage the same authentication channel, rather than forcing every connection to individually authenticate
+- Could support pre-authentication + pre-population (subject to scale concerns)
+
+*Cons*
+- More expensive per-packet to look up an additional map using packet fields
+- Additional memory overhead (~N MB)
+
+###### Implementation details
+Q: How big to make the auth map in BPF? Depends on the key type and target number of peers in the cluster. Assuming ~16B for IPv4 + ~40B for IPv6, => 56 (round to 64B), 64K peers could be supported in ~4MB of memory. This comes down to 1.5MB or less if keyed by identity.
```

Comment:
Another case where we've defined the key as 3x`__u32` now, but this rough math makes other assumptions.

Reply:
(implementation details: this is probably somehow padded/aligned on some word boundary so common case for 3x`__u32` is probably equivalent to 4x`__u32`, ie ~16B, so about 1MB as the last sentence implies)